### ISSUE TITLE
AddSymbolicGradients segfaults internally C++ API

### ISSUE THREAD

**Issue by Proch92** [tensorflow/tensorflow#24670]
**System information**
- Have I written custom code:
- Linux Ubuntu 16.04
- TensorFlow installed from source:
- TensorFlow version: r1.13 #24493 
- CPU only
- Bazel version: Build label: 0.21.0
- GCC/Compiler version: gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
- memory: 6GB

**Describe the current behavior**
AddSymbolicGradients call results in a segmentation fault.

**Describe the expected behavior**
AddSymbolicGradients should return grads or fail safely within TF_CHECK_OK

**Code to reproduce the issue**
```
auto input_slices = Split(scope, 1, x, window_size);

auto initial_state = Fill(scope, {batch_size, state_size}, 0);

vector<Output> states;
states.reserve(window_size+1);
states.push_back(initial_state);

for (int i=0; i!=window_size; i++) {
    auto concat = Concat(scope, InputList(initializer_list<Input>{input_slices[i], states[i]}), 1);
    auto new_state = Tanh(scope, Add(scope, MatMul(scope, concat, w_rnn), b_rnn));
    states.push_back(new_state);
}

// dense output
auto out = Tanh(scope, Add(scope, MatMul(scope, states[window_size], w_dense), b_dense));

// loss function
auto loss = ReduceMean(scope, Square(scope, Sub(scope, out, y)), {0, 1});

vector<Output> grad_outputs;
TF_CHECK_OK(AddSymbolicGradients(scope, {loss}, {w_rnn, w_dense, b_rnn, b_dense}, &grad_outputs));
```

**Other info / logs**
gdb where command output:

> (gdb) where
> #0  0x00005555559b45aa in tensorflow::(anonymous namespace)::SymbolicGradientBuilder::Initialize() ()
> #1  0x00005555559b6f1c in tensorflow::AddSymbolicGradients(tensorflow::Scope const&, std::vector<tensorflow::Output, std::allocator<tensorflow::O
> utput> > const&, std::vector<tensorflow::Output, std::allocator<tensorflow::Output> > const&, std::vector<tensorflow::Output, std::allocator<tens
> orflow::Output> > const&, std::vector<tensorflow::Output, std::allocator<tensorflow::Output> >*) ()
> #2  0x00005555559b9bb2 in tensorflow::AddSymbolicGradients(tensorflow::Scope const&, std::vector<tensorflow::Output, std::allocator<tensorflow::O
> utput> > const&, std::vector<tensorflow::Output, std::allocator<tensorflow::Output> > const&, std::vector<tensorflow::Output, std::allocator<tens
> orflow::Output> >*) ()
> #3  0x000055555597ce6c in Rnn::train(Dataset, int, int, int) ()
> #4  0x00005555558410d6 in main ()

Variables scope nor vectors should be the issue here. i've tried to unfold the loop without luck. Any hint?

**Comment by msymp** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by ptzenov** [comment]
I know the topic is closed, but since I ran into the same problem, I figured I should mention my insights  here. My guess is that the bug is related to the `ConcatGrad` operation, which gives an empty result when computing gradients with respect to a variable which is one of the constituents of the `Concat` operation. This could lead to seg-faults, when propagating your gradient to further nodes. 

Consider the following examle:
```
auto x1 = Variable(scope.WithOpName("x1"), {2, 1}, DT_FLOAT);                                                            
auto x2 = Variable(scope.WithOpName("x2"), {2, 1}, DT_FLOAT);                                                            
auto x = Concat(scope.WithOpName("x"), InputList(std::initializer_list<Output>{x1, x2}), 0);                             
auto y = MatMul(scope.WithOpName("y"), x, x, MatMul::TransposeA(true));                                                                                                                                                                          
std::vector<Output> grad;                                                                                                                                                                          
AddSymbolicGradients(scope.WithOpName("gradient"), {y}, {x1}, &grad);       // grad will be empty                                              
```                          
In this case, `tf` cannot   compute dy/dx1 (thus `grad` will be an empty list). As a work-around,  either compute dy/dx or  (if you need the derivative of `y` with respect to `x1`) eliminate the `Concat` operation whatsoever, by representing the matmul operation explicitly, e.g. in my example u add two `MatMul` operations and add them to get the end result.
